dlib/test/image.cpp

```diff
@@ -1987,6 +1987,222 @@ namespace
         }
     }
 
+    template<typename interpolation_type = interpolate_bilinear>
+    void test_resize_image_with_interpolation()
+    {
+        {
+            matrix<unsigned char> img_s(2, 2);
+            matrix<unsigned char> img_d(3, 3);
+
+            img_s(0, 0) = 0;
+            img_s(0, 1) = 100;
+            img_s(1, 0) = 100;
+            img_s(1, 1) = 100;
+
+            resize_image(img_s, img_d, interpolation_type());
+            DLIB_TEST((img_d(0, 0) == 0));
+            DLIB_TEST((img_d(0, 1) == 50));
+            DLIB_TEST((img_d(1, 2) == 100));
+            DLIB_TEST((img_d(2, 2) == 100));
+        }
+
+        {
+            matrix<rgb_pixel> img_s(2, 2);
+            matrix<rgb_pixel> img_d(3, 3);
+
+            img_s(0, 0) = { 0, 0, 0 };
+            img_s(0, 1) = { 10, 20, 30 };
+            img_s(1, 0) = { 10, 20, 30 };
+            img_s(1, 1) = { 10, 20, 30 };
+
+            auto same = [](rgb_pixel p1, rgb_pixel p2)
+            {
+                return std::abs(p1.red - p2.red) + std::abs(p1.green - p2.green) + std::abs(p1.blue - p2.blue) == 0;
+            };
+
+            resize_image(img_s, img_d, interpolation_type());
+            DLIB_TEST((same(img_d(0, 0), { 0, 0, 0 })));
+            DLIB_TEST((same(img_d(0, 1), { 5, 10, 15 })));
+            DLIB_TEST((same(img_d(1, 2), { 10, 20, 30 })));
+            DLIB_TEST((same(img_d(2, 2), { 10, 20, 30 })));
+        }
+
+        {
+            matrix<lab_pixel> img_s(2, 2);
+            matrix<lab_pixel> img_d(3, 3);
+
+            img_s(0, 0) = { 0, 0, 0 };
+            img_s(0, 1) = { 100, 20, 30 };
+            img_s(1, 0) = { 100, 20, 30 };
+            img_s(1, 1) = { 100, 20, 30 };
+
+            auto same = [](lab_pixel p1, lab_pixel p2)
+            {
+                return std::abs(p1.l - p2.l) + std::abs(p1.a - p2.a) + std::abs(p1.b - p2.b) == 0;
+            };
+
+            resize_image(img_s, img_d, interpolation_type());
+            DLIB_TEST((same(img_d(0, 0), { 0, 0, 0 })));
+            DLIB_TEST((same(img_d(0, 1), { 50, 10, 15 })));
+            DLIB_TEST((same(img_d(1, 2), { 100, 20, 30 })));
+            DLIB_TEST((same(img_d(2, 2), { 100, 20, 30 })));
+        }
+
+    }
+    
+    template<typename interpolation_type = interpolate_bilinear>
+    void test_null_rotate_image_with_interpolation()
+    {
+        {
+            matrix<unsigned char> img_s(3, 3);
+            matrix<unsigned char> img_d;
+
+            img_s(0, 0) = 0;
+            img_s(0, 1) = 100;
+            img_s(0, 2) = 100;
+            img_s(1, 0) = 100;
+            img_s(1, 1) = 100;
+            img_s(1, 2) = 100;
+            img_s(2, 0) = 100;
+            img_s(2, 1) = 100;
+            img_s(2, 2) = 100;
+
+            rotate_image(img_s, img_d, 0, interpolation_type());
+            DLIB_TEST((img_d(0, 0) == 0));
+            DLIB_TEST((img_d(0, 1) == 100));
+            DLIB_TEST((img_d(1, 0) == 100));
+            DLIB_TEST((img_d(1, 1) == 100));
```

Comment:
Seems like this should be `DLIB_TEST(img_d == img_s)`.  Here and elsewhere, if that's what's intended use operator== to compare the objects rather than comparing subcomponents.

Reply:
In fact, they are not equal while preferably they should be. I think there is a problem with how interpolate_bilinear works:
- on edges
- for sub-pixel accuracy.

Right now it works like this (if we imagine white is a huge picture interpolated out of original `img`):
![image](https://user-images.githubusercontent.com/8502965/82901176-de629900-9f5d-11ea-86fb-f9fa691cd0fc.png)
The value in `img` is from upper top corner of pixel in target higher resolution image. Hence on right and bottom edge, there is no much information, only extrapolation possible. This will result it right and bottom edge having some artifact, while left and top being "perfect".

Moreover, since in code there is:
```
// if the interpolation goes outside img 
if (!(left >= 0 && top >= 0 && right < img.nc() && bottom < img.nr()))
    return false;
```
in fact there won't be extrapolation happening. So the artifact won't be "same color in each row/column", but "no color setting". Worse, it affects also last values in each row/column (red cross). Hence rotating by 0 degrees produces not equal image :(

Other solution could be used where color value in `img` (blue) is in the middle of target high resolution image (white), plus clamping coordinates of 4 pixels used for interpolation (green):
![image](https://user-images.githubusercontent.com/8502965/82901953-fbe43280-9f5e-11ea-84c4-b847c7850fff.png)
This would have a problem though as user would have to be aware of it and provide half pixel offset in necessary.

Maybe just clamping coordinates would be enough for now? So that interpolation uses closest existing value and always "succeeds"? It will still have asymmetrical artifact without cautious use, but at least someone forgetting to provide half pixel offset won't get wrong blurry output.
Downside with clamping might be also that it's additional condition, although can write it in a way it hints to CPU which branch will be usually taken.

Reply:
Hmm, adding just clamping results in fail in `test_extract_image_chips()`, likely this happens:
![image](https://user-images.githubusercontent.com/8502965/82915479-12938500-9f71-11ea-8116-b7b6e66751c8.png)
Previously those red lines outside of `img` wouldn't exist...
So there is somewhat more backward compatibility breakage than I would expect. Not sure how to proceed.

Reply:
Ah, well, don't worry about it.  I was just noting that the test code could be simpler, but border effects make it not so.  I would rather keep things the same, as changes to this behavior could break user code that depends on it being the way it is.  In particular, machine learning models trained downstream of this could easily depend on the particular details.  Moreover, in many appliations border effects are not important and the compute cost to do someting neater isn't worth it.